### PR TITLE
Admin order cycles performance

### PR DESCRIPTION
**Pull request by Matt-Yorkley**
#### What? Why?

Related to #3858 

I did some investigation and managed to find a couple of places where there were quick wins. I managed to improve a couple of endpoints related to Order Cycles index and edit pages.

Endpoint before: 548 queries, 2.9 seconds.
Screenshot from 2019-05-28 23-58-21

Endpoint after: 42 queries, 0.4 seconds.
Screenshot from 2019-05-29 00-00-56

Endpoint before: 2048 queries, 12.2 seconds.
Screenshot from 2019-05-29 00-53-29

Endpoint before: 952 queries, 7.3 seconds.
enpoint_after

The remaining 952 queries on this page are mostly this, repeating:

```ruby
lib/open_food_network/variant_and_line_item_naming.rb:13:in `options_text'
lib/open_food_network/variant_and_line_item_naming.rb:42:in `unit_to_display'
lib/open_food_network/variant_and_line_item_naming.rb:30:in `full_name'
app/serializers/api/admin/for_order_cycle/supplied_product_serializer.rb:22:in `block in variants'
app/serializers/api/admin/for_order_cycle/supplied_product_serializer.rb:22:in `variants'
app/controllers/admin/enterprises_controller.rb:100:in `block (2 levels) in for_order_cycle'
app/controllers/admin/enterprises_controller.rb:98:in `for_order_cycle'

SELECT "spree_option_values".* FROM "spree_option_values" INNER JOIN "spree_option_types" 
ON "spree_option_types"."id" = "spree_option_values"."option_type_id" 
INNER JOIN "spree_option_values_variants" 
ON "spree_option_values"."id" = "spree_option_values_variants"."option_value_id" 
WHERE "spree_option_values_variants"."variant_id" = 203347 
ORDER BY spree_option_types.position asc;
```
There's obviously more N+1 problems here, but I didn't manage to crack this bit. :point_up: If anyone else fancies a go, be my guest. 

I think if we want more improvements on order cycles we'll need some big changes. I think we currently preload basically everything into Angular objects. We can move to doing quick API calls for some of the UI parts, like adding a product to the OC or selecting an enterprise.

#### What should we test?

Admin order cycle index and edit pages should work as before, and load faster.

#### Release notes

Improved performance on admin Order Cycle pages

Changelog Category: Changed

### CODE REVIEW

**Comment by RachL** [comment]
@Matt-Yorkley is this closing #3858 or is further investigation needed?

**Comment by Matt-Yorkley** [comment]
> @Matt-Yorkley is this closing #3858 or is further investigation needed?

It's not closing the spike yet. I'll write up a few things I found and make issues for them if needed.

File: app/controllers/admin/enterprises_controller.rb [hunk 1]
```diff
@@ -142,7 +142,15 @@ def collection
         @order_cycle = OrderCycle.find_by_id(params[:order_cycle_id]) if params[:order_cycle_id]
         coordinator = Enterprise.find_by_id(params[:coordinator_id]) if params[:coordinator_id]
         @order_cycle = OrderCycle.new(coordinator: coordinator) if @order_cycle.nil? && coordinator.present?
-        OpenFoodNetwork::OrderCyclePermissions.new(spree_current_user, @order_cycle).visible_enterprises
+
+        enterprises = OpenFoodNetwork::OrderCyclePermissions.new(spree_current_user, @order_cycle)
+          .visible_enterprises
+
+        unless enterprises.empty?
```
Comment (sauloperez) [inline app/controllers/admin/enterprises_controller.rb]: why can't we just concat this with the `includes` below. Isn't it an ActiveRecord::Relation? It looks like it is by looking at line 166.

Comment (Matt-Yorkley) [reply app/controllers/admin/enterprises_controller.rb]: There were lots of spec failures when I tried that. In some of the tests it's an empty array.

**Comment by lin-d-hop** [comment]
Everything is working as expected:
https://docs.google.com/document/d/1RT7GsSCOp8UWFsMznMOz8by6p55o1KmSGBWd5xMc8Cs

Note I did not create enough OCs to have an impact on performance. I did not test if there were performance improvements - only that OC pages all continue to work as expected.

@matt-yorkley - Should we do some benchmarking as this is deployed to prod?

**Comment by sauloperez** [comment]
> @Matt-Yorkley - Should we do some benchmarking as this is deployed to prod?

Yes! the app behaves quite differently in production compared to development and we need to get accurate estimates as we move forward with the epic.

**Pull request merged**